### PR TITLE
nixos/scx: init module

### PR DESCRIPTION
**Pull request by JohnRTitor**
This adds a `services.scx.enable` option to enable sched-ext schedulers.

Requires a kernel with sched-ext enabled (6.12+) or a kernel with the patchset. As of this PR, in Nixpkgs only `linuxPackages_testing` is compatible.

Since kernel upgrades are also backported, this can be backported as well.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux

- [x] Tested, as applicable:
  - With my own configuration
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by s0me1newithhand7s** [comment]
impressive! that would be so nice to see in nixpkgs, i'll wait for this pr to be merged, definitely a needed one.

**Comment by JohnRTitor** [comment]
Built NixOS manual.
image
image

**Comment by emilazy** [comment]
Sorry, I don’t really know anything about this. The kernel maintainers would probably be in a better position to review?

File: nixos/modules/services/scheduling/scx.nix
Comment (K900) [inline nixos/modules/services/scheduling/scx.nix:98]: Unnecessary.

File: nixos/modules/services/scheduling/scx.nix
Comment (K900) [inline nixos/modules/services/scheduling/scx.nix:83]: You can assert on the kernel config here, see `requiredKernelFeatures`.

**Review by K900 — DISMISSED** [review]
Also, please proofread the text, there's a lot of typos and questionable wording in the documentation.

File: nixos/modules/services/scheduling/scx.nix
Comment (JohnRTitor) [reply nixos/modules/services/scheduling/scx.nix:104]: Sure, I based this off https://github.com/sched-ext/scx/blob/main/services/systemd/scx.service where it was like that.

File: nixos/modules/services/scheduling/scx.nix
Comment (JohnRTitor) [reply nixos/modules/services/scheduling/scx.nix:83]: I am not finding any result when I look up `requiredKernelFeatures` in across Nixpkgs, perhaps you mean `requiredKernelConfig`?

I suppose this would work?
```
    system.requiredKernelConfig = with config.lib.kernelConfig; [
      (isYes "SCHED_CLASS_EXT")
      (isYes "BPF")
    ];
```

Comment (K900) [reply nixos/modules/services/scheduling/scx.nix:83]: Yes, that.

Comment (JohnRTitor) [reply nixos/modules/services/scheduling/scx.nix:81]: Done. Purposefully excluded the PAHOLE options.

File: nixos/modules/services/scheduling/scx.nix
Comment (JohnRTitor) [reply nixos/modules/services/scheduling/scx.nix:109]: That should do it. Thanks.

**Comment by JohnRTitor** [comment]
I copyedited the text. I hope it's okay, Please suggest if you want to make a change.

File: nixos/modules/services/scheduling/scx.nix
Comment (K900) [reply nixos/modules/services/scheduling/scx.nix:109]: What should do what? The point is that SCHED_CLASS_EXT depends on all of those things already, so if it is enabled, then so must be all of its dependencies.

File: nixos/modules/services/scheduling/scx.nix
Comment (K900) [inline nixos/modules/services/scheduling/scx.nix:85]: Does this actually make sense? When would you want this?

File: nixos/modules/services/scheduling/scx.nix
Comment (JohnRTitor) [reply nixos/modules/services/scheduling/scx.nix:109]: I thought you wanted me to only keep the dependencies. So I removed the rest.

Anyways, I see how it is. But keeping dependencies in assertion wouldn't hurt, right?

Comment (K900) [reply nixos/modules/services/scheduling/scx.nix:109]: What would be the reason to keep them?

File: nixos/modules/services/scheduling/scx.nix
Comment (JohnRTitor) [reply nixos/modules/services/scheduling/scx.nix:85]: So just `SCX scheduler`? or do you want me to put actual description in there

File: nixos/modules/services/scheduling/scx.nix
Comment (JohnRTitor) [reply nixos/modules/services/scheduling/scx.nix:109]: None.

File: nixos/modules/services/scheduling/scx.nix
Comment (K900) [reply nixos/modules/services/scheduling/scx.nix:85]: Sorry, wrong line, I meant adding it to `systemPackages`.

Comment (JohnRTitor) [reply nixos/modules/services/scheduling/scx.nix:85]: In that case, my defense would be:

Let's say I am testing another scheduler `scx_bpfland`, I will stop this service (my permanent scheduler) temporarily and manually run `sudo scx_bpfland`. Not having it in `environment.systemPackages` will not have it in PATH without a `nix-shell`.

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (K900) [reply nixos/doc/manual/release-notes/rl-2411.section.md:30]: "available _as_ `services.scx`", and code block the option name.

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (K900) [inline nixos/doc/manual/release-notes/rl-2411.section.md:31]: "ramp up" -> "improve", but maybe just remove that entire sentence unless you have strong data to back it up.

"Requires ~~running~~ a Linux (capitalization) kernel ~~of~~ version 6.12 or later."

File: nixos/modules/services/scheduling/scx.nix
Comment (K900) [inline nixos/modules/services/scheduling/scx.nix:19]: How does this documentation actually help?

File: nixos/modules/services/scheduling/scx.nix
Comment (K900) [inline nixos/modules/services/scheduling/scx.nix:15]: Again, claims of "improve system performance" are likely unfounded in the general case.

File: nixos/modules/services/scheduling/scx.nix
Comment (K900) [inline nixos/modules/services/scheduling/scx.nix:25]: Use `mkPackageOption`.

File: nixos/modules/services/scheduling/scx.nix
Comment (K900) [reply nixos/modules/services/scheduling/scx.nix:85]: OK, that's not unreasonable.

File: nixos/modules/services/scheduling/scx.nix
Comment (K900) [inline nixos/modules/services/scheduling/scx.nix:89]: Nit: this attrset can be a single line.

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (emilazy) [reply nixos/doc/manual/release-notes/rl-2411.section.md:30]: You can just do `[](#opt-…)` and it will be filled out automatically, AIUI.

File: nixos/modules/services/scheduling/scx.nix
Comment (JohnRTitor) [reply nixos/modules/services/scheduling/scx.nix:25]: This was causing some issues. Can't override the description apparently.

And using it like:

```nix
lib.mkPackageOption pkgs [ "scx" "full" ] { }
```
the description to be: "which full package to use"

File: nixos/modules/services/scheduling/scx.nix
Comment (JohnRTitor) [reply nixos/modules/services/scheduling/scx.nix:19]: I added in the hopes of a clueless user finding these:

Oh I don't have the required kernel, let's see oh I need to build the kernel with these options.

File: nixos/modules/services/scheduling/scx.nix
Comment (JohnRTitor) [reply nixos/modules/services/scheduling/scx.nix:15]: Will remove it.

File: nixos/modules/services/scheduling/scx.nix
Comment (K900) [reply nixos/modules/services/scheduling/scx.nix:19]: A user clueless enough to see an assertion about their kernel config and not understand what it means should not be allowed near custom kernels for their own safety.

File: nixos/modules/services/scheduling/scx.nix
Comment (K900) [reply nixos/modules/services/scheduling/scx.nix:25]: OK.

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (JohnRTitor) [reply nixos/doc/manual/release-notes/rl-2411.section.md:31]: > but maybe just remove that entire sentence unless you have strong data to back it up.

I don't have the patience to generate graphs nor I will look for benchmarks. But I observed low to none lag when browsing my system, watching my videos, all while compiling a kernel for instance. Again, this was my on my own setup.

Upstream claims this though, with a [video reference](https://github.com/sched-ext/scx?tab=readme-ov-file#sched_ext-schedulers-and-tools).

File: nixos/modules/services/scheduling/scx.nix
Comment (emilazy) [reply nixos/modules/services/scheduling/scx.nix:25]: You can do `lib.mkPackageOption pkgs [ "scx" "full" ] // { … }`.

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (K900) [reply nixos/doc/manual/release-notes/rl-2411.section.md:31]: I suggest we err on the side of caution here, and don't promise anything specific, especially on a feature that's largely still experimental.

File: nixos/modules/services/scheduling/scx.nix
Comment (JohnRTitor) [reply nixos/modules/services/scheduling/scx.nix:25]: I tried it before. NixOS manual fails to build.
```nix
lib.mkPackageOption pkgs [ "scx" "full" ] // {
      description = ''
        `scx` package to use. ....
      '';
    };
```

Comment (emilazy) [reply nixos/modules/services/scheduling/scx.nix:25]: Ah, sorry: the correct way is `lib.mkPackageOption pkgs "friendly name" { default = [ "scx" "full" ]; }`. Anyway, doesn’t matter.

Comment (JohnRTitor) [reply nixos/modules/services/scheduling/scx.nix:25]: This just makes it more complex for no reason IMO.

Comment (emilazy) [reply nixos/modules/services/scheduling/scx.nix:25]: It would save several lines of code and docs here, but sure.

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (JohnRTitor) [reply nixos/doc/manual/release-notes/rl-2411.section.md]: > You can just do `[](#opt-…)` and it will be filled out automatically, AIUI.

Even on GitHub?

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (K900) [inline nixos/doc/manual/release-notes/rl-2411.section.md:31]: Requires ~~a~~ Linux kernel ~~of~~ version 6.12 or later.

File: nixos/modules/services/scheduling/scx.nix
Comment (K900) [inline nixos/modules/services/scheduling/scx.nix:103]: Is it a bool or a tristate?

Comment (JohnRTitor) [reply nixos/modules/services/scheduling/scx.nix:103]: Bool as per https://cateee.net/lkddb/web-lkddb/SCHED_CLASS_EXT.html

Comment (K900) [reply nixos/modules/services/scheduling/scx.nix:103]: OK

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (JohnRTitor) [reply nixos/doc/manual/release-notes/rl-2411.section.md]: > I'd rather hold this off until kernel 6.12 is stable, at least.

Won't this hard fail on kernels that are not supported though?

I don't think holding this would contribute to its testing/stability. Letting it run on nixos-unstable will attract more bug reports if something is wrong. And once something is merged/approved on upstream kernel, there is little to no change that it will be reverted.

**Comment by JohnRTitor** [comment]
Since 6.12 have landed, can we get this going? @K900

**Comment by K900** [comment]
It has not, in fact, landed (yet).

**Comment by JohnRTitor** [comment]
Nice, looks like https://github.com/NixOS/nixpkgs/pull/356881 got merged.

File: nixos/modules/services/scheduling/scx.nix
Comment (Mic92) [reply nixos/modules/services/scheduling/scx.nix:103]: This option has no effect for quite a while now since we now have "features" rather than kerne config. So there is actually no point in adding this. The other instances of this config can be dropped from nixpkgs as well.

Comment (JohnRTitor) [reply nixos/modules/services/scheduling/scx.nix:103]: So what would be the appropriate way to check if the kernel has SCX support?

https://github.com/NixOS/nixpkgs/pull/352300/files#diff-999da5c78431b99032ddb89d748beb2fcce2d46ea5a95fb5195cdec042dde310R86
```
systemd unitConfig.ConditionPathIsDirectory = "/sys/kernel/sched_ext";
```
this prevents the service from starting under an unsupported kernel, but it's a runtime check not a config eval time check. Would this be enough?

Comment (Mic92) [reply nixos/modules/services/scheduling/scx.nix:103]: Is it enabled in all new kernels? We could check the kernel version for now.

Comment (JohnRTitor) [reply nixos/modules/services/scheduling/scx.nix]: Yup, changed to:

```nix
    assertions = [
      {
        assertion = lib.versionAtLeast config.boot.kernelPackages.kernel.version "6.12";
        message = "SCX is only supported on kernel version >= 6.12.";
      }
    ];
```

**Comment by JohnRTitor** [comment]
Let's not block any further. The condition to merge this after 6.12 reached nixpkgs have been met. I see no review comments blocking this for any other reason.

File: nixos/modules/services/scheduling/scx.nix
Comment (oxalica) [inline nixos/modules/services/scheduling/scx.nix:95]: This is already the default value according [systemd.exec(5)](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#StandardOutput=).

**Pull request merged**
